Implement accessibility using AccessKit

geom3trik opened the pull request:
This PR implements integration with [AccessKit](https://github.com/AccessKit/accesskit), which provides an accessibility API for users of screen readers.

### Accessible Views:
- [x] Label
- [x] Button
- [x] Checkbox
- [x] Slider
- [x] RadioButton
- [x] Textbox

## TODO
- [ ] Make all views accessible where possible
- [ ] Make all examples accessible where possible
- [x] Figure out an API for `labelled_by` (could borrow from describing api of label)

Currently this only works on Windows until AccessKit adds support for other operating systems.

geom3trik commented:
There's a design question I wouldn't mind getting some opinions on. Because of the way accesskit does tree updates (updates to the accessibility tree), all the information required for a node needs to be available in vizia or should be derived from information available in vizia.

For example, we can't just store the role of a view in the accessibility tree because when we do a tree update we need to fill in the role property of the node or it will remove the role. So I'm storing the role in the context, keyed on the entity. However, there are some node properties that end up getting duplicated within vizia if we do everything this way.

An example would be the `Slider` view which binds to some value. A copy of that value is needed for the accessibility tree so currently I'm storing an aditional `value` in context, also keyed on entity. Is this an okay solution even though it duplicates a little bit of state?

Another potential solution could be to add another method to the `View` trait, something like `accessibility()` which gets called every frame and the user can fetch the node data for their view, mutate the value property, and the push this as a tree update to be sent to the accesskit adapter. This would remove the duplication of state but at the cost of a more complex API for the user making custom views.

Open to any and all opinions on this and any solutions I haven't thought of.

rhelmot commented:
I would need to see the implementation but I'm fine with data duplication as long as it's hard for it to be out-of-sync in a bad way. Anything we can do to make accessibility easier to program for (not having a dedicated accessibility method) will pay off in the end imo.

geom3trik commented:
> I would need to see the implementation but I'm fine with data duplication as long as it's hard for it to be out-of-sync in a bad way. Anything we can do to make accessibility easier to program for (not having a dedicated accessibility method) will pay off in the end imo.

I think I would agree. So far I've managed to implement accessibility for the views which can be made accessible (with what is currently available in accesskit) without the need for an accessibility method. This does require some duplication of state but it's easy enough to keep it in sync. It also requires some special logic just for textboxes but I'm using the role to selectively apply this logic within an accessibility system which runs on all of the views.

rhelmot commented on crates/vizia_winit/src/application.rs:
```diff
@@ -30,9 +36,29 @@ use winit::{
+pub enum UserEvent {
+    Event(Event),
+    #[cfg(not(target_arch = "wasm32"))]
+    AccessKitActionRequest(accesskit_winit::ActionRequestEvent),
```
If this is a winit limitation, can we avoid changing the public interfaces and avoid the into() calls?

geom3trik replied:
This is because accesskit is utilising winit custom events to receive action requests from a screen reader. But we're already using custom events to send vizia events between threads with the event loop proxy. So this enum is required so we can keep both functionality. I'm open to suggestions on API improvements if something can be done.

geom3trik commented on crates/vizia_core/src/style/mod.rs:
```diff
@@ -95,6 +96,19 @@ pub struct Style {
+    pub live: SparseSet<Live>,
+    pub labelled_by: SparseSet<Entity>,
+    pub hidden: SparseSet<bool>,
+    // TODO: Should we store these or somehow derive them from app state?
```
But then we would still need to store this state somewhere so that we can build the accesskit node. Note that entities which don't have these properties don't have a space in the sparse set.

geom3trik commented on crates/vizia_core/src/systems/accessibility.rs:
```diff
@@ -0,0 +1,224 @@
+// Updates node properties from view properties
+// Should be run after layout so that things like bounding box are correct
+// This system doesn't change the structure of the accessibility tree as this is done when views are built/removed
+// TODO: Change this to incrementally update nodes when required instead of updating all nodes every frame
```
I haven't profiled it, but accesskit is designed for the case where the nodes need to be updated every frame to accommodate immediate mode GUIs. However, since we do know what changes it probably wouldn't be difficult to update only the nodes that changed. It's just whether or not we want to block on that.

geom3trik commented on crates/vizia_core/src/systems/accessibility.rs:
```diff
@@ -0,0 +1,224 @@
+        // Here we need to construct the correct text edit nodes for each wrapped line of text
+        if let Some(role) = cx.style.roles.get(entity) {
+            if *role == Role::TextField {
+                // This is a dirty hack because we need the bounds of the inner inner text content
```
It would be improved somewhat, but it would be the same code in a different place.

geom3trik commented on crates/vizia_core/src/systems/accessibility.rs:
```diff
@@ -0,0 +1,224 @@
+
+                        // FIXME: The screen reader gets out of sync with the cursor due to affinity
+
+                        // TODO: Might need to append any spaces that were stripped during layout. This can be done by
```
I'll have to double check but it seemed to me that lines which wrap can sometimes have the trailing space removed when you query the text in the line, which is done by taking the slice of text from the hard wrapped line (what cosmic gives you in a run) from the start glyph to the end glyph of that run (because the glyphs take in to account soft wrap). Most of the pain of this stems from how cosmic keeps track of lines of text. Accesskit needs a node per line, whether it's soft or hard wrapped.

geom3trik commented on crates/vizia_core/src/view.rs:
```diff
@@ -28,6 +32,21 @@ pub trait View: 'static + Sized {
+        let children =
+            parent_id.child_iter(&cx.tree).map(|entity| entity.accesskit_id()).collect::<Vec<_>>();
+
+        cx.tree_updates.push(TreeUpdate {
```
Potentially... this is updating the accessibility tree when a view is created, which occurs when `build` is called. More specifically this is updating the structure of the tree rather than the values on the nodes.

geom3trik commented on crates/vizia_core/src/views/textbox.rs:
```diff
@@ -756,6 +770,19 @@ where
+                data: Some(ActionData::SetTextSelection(_selection)),
+            }) => {
+                // println!("Select some text: {:?}", selection);
+                // TODO: Handle selection action request
```
Hmm probably should. I'm just avoiding the horrible mess that entails because I'll have to iterate the line glyphs to figure it out...

geom3trik commented:
> Revisiting the conversation we had earlier in this thread about view methods vs style entries, I think the current API (style entries) does not provide an improvement in usability over view-method-based getters. My hope was that we would be able to populate some of these fields automatically, but as it stands you would need to add a line to the constructor to set or bind the appropriate property, and in my mind this is just as hard to remember to do as implementing a separate method on the View trait, perhaps even harder because IDEs can give you suggestions for methods to implement. We could even make the accessibility methods required implementations to force the user to consider them up front.

I'm not sure. A lot of accessibility properties can be added as modifiers on a view, and I think that's in keeping with how a lot of other things in vizia work. But then some things require some form of specialisation, such as the textbox (as seen in the accessibility system which currently switches on role to accommodate that). Another thought I had is that custom views that implement their own rendering need some way to construct an accessibility subtree, and for that a method on `View` may be the better way to go.

geom3trik commented:
Okay I think this is ready for another review. I've updated to the latest version of accesskit, which changes the way nodes are built by providing a `NodeBuilder`. I've introduced 3 main new parts since last time:
1. An `AccessNode` which wraps a `NodeBuilder`, a `NodeId`, and a list of children that can be added to a node which don't derive from a view (e.g. text lines).
2. An `AccessContext` to be used with...
3. An `accessbility` method on `View` so that views can customise their accessibility properies more so than what is possible with the accessibility modifiers. So fa this is used with the `Textbox` view for generating accessibility nodes for each text line, and with the `Slider` for setting the min/max value and step.

There's 3 main things to sort now but I don't know if they're neccessarily blocking:
- [ ] The accessibility system which updates nodes based on style/accessibility properties could be configured to run incrementally on just the views that require it.
- [ ] The textbox is missing setting the selection from an action. This is tricky because of how cosmic-text stores layout lines.
- [ ] Multiline text isn't being read properly when the cursor is at the start of a soft wrapped line. I think this something to do with the cursor affinity but I'm not yet sure how to fix this.

geom3trik commented:
Okay, I'm pretty eager to get this merged so I'm going to push the final 2 issues to follow-up PRs. The first is converting an accesskit text selection to something cosmic understands. I've started to do this but I need to figure out how to actually trigger this with a screen reader to test it and continue. The second is the incrementally updating views when accessinility properties change. This isn't difficult but it has a large surface area due to how many different things can effect accessibility nodes, therefore I think it's fine to push this to a follow-up PR.

Everything is working correctly with the NVDA screen reader but there are some bugs with the Narrator screen reader built into Winows, specifically regarding multiline text. I'm investigating this, but since NVDA is far more common (I've been told) and is working correctly I don't see this as a blocker to get this merged.

rhelmot commented:
wooooooooooooooooooo
